text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,877 @@
+   * CloudFront distribution access. If the imported bucket was previously configured with OAI and being
+   * migrated to OAC, the OAI policy statements must be removed manually.
+   * @default false
+   */
+  readonly overrideImportedBucketPolicy?: boolean;
+
+  /**
+   * The level of permissions granted in the bucket policy and key policy (if applicable)
+   * to the CloudFront distribution. This property only applies to OAC (not OAI).
+   * @default AccessLevel.READ
+   */
+  readonly originAccessLevels?: AccessLevel[];
+}
+
+/**
+ * The types of permissions to grant OAC access to th S3 origin
+ */
+export enum AccessLevel {
+  /**
+   * Grants 's3:GetObject' permission to OAC
+   */
+  READ = 'READ',
+  /**
+   * Grants 's3:PutObject' permission to OAC
+   */
+  WRITE = 'WRITE',
+  /**
+   * Grants 's3:DeleteObject' permission to OAC
+   */
+  DELETE = 'DELETE',
+}
+
+
+export class S3Origin implements cloudfront.IOrigin {
+  private readonly origin: cloudfront.IOrigin;
+
+  constructor(bucket: s3.IBucket, props: S3OriginProps = {}) {
+    if (props.originAccessControl && props.originAccessIdentity) {
+      throw new Error('Only one of originAccessControl or originAccessIdentity can be specified for an origin.');
+    }
+
+    if (bucket.isWebsite) {
+      this.origin = new HttpOrigin(bucket.bucketWebsiteDomainName, {
+        protocolPolicy: cloudfront.OriginProtocolPolicy.HTTP_ONLY, // S3 only supports HTTP for website buckets
+        ...props,
+      });
+    } else if (props.originAccessControl) {
+      this.origin = S3BucketOrigin.withAccessControl(bucket, props);
+    } else if (props.originAccessIdentity) {
+      this.origin = S3BucketOrigin.withAccessIdentity(bucket, props);
+    } else {
+      this.origin = FeatureFlags.of(bucket.stack)
+        .isEnabled(cxapi.CLOUDFRONT_USE_ORIGIN_ACCESS_CONTROL_BY_DEFAULT) ?
+        S3BucketOrigin.withAccessControl(bucket, props) :
+        S3BucketOrigin.withAccessIdentity(bucket, props);
+    }
+  }
+
+  public bind(scope: Construct, options: cloudfront.OriginBindOptions): cloudfront.OriginBindConfig {
+    return this.origin.bind(scope, options);
+  }
+}
+```
+
+#### `Distribution` construct modifications
+
+In the `addOrigin()` method of `Distribution`, we will need to pass the `distributionId` to `origin.bind()` to specify the condition in the policy statement.
+
+```ts
+  private addOrigin(origin: IOrigin, isFailoverOrigin: boolean = false): string {
+    const ORIGIN_ID_MAX_LENGTH = 128;
+
+    const existingOrigin = this.boundOrigins.find(boundOrigin => boundOrigin.origin === origin);
+    if (existingOrigin) {
+      return existingOrigin.originGroupId ?? existingOrigin.originId;
+    } else {
+      ...
+      const distributionId = this.distributionId;
+      const originBindConfig = origin.bind(scope, { originId: generatedId, distributionId: Lazy.string({ produce: () => this.distributionId }) });
```

Comment:
For context, Lazys are for values that you can't know in the constructor (this is at least part of their purpose, they might be useful for more). Imagine you had a list property, like say a list of notification ARNs to send notifications to.

There's two ways a user could specify these ARNs: in the constructor, or through a method, `addNotificationArn()`. 

This creates a problem; in the constructor, you have to call the `Cfn*` L1 constructor, which needs the value of `notifitcationArns`. If you just pass the `props.notificationArns` directly, then you lose any props that were configured with the `addNotificationArn()` method, since that can't be called on the object until the constructor is finished.

How do you pass both sources to the L1? You use a Lazy, which will let you add all the sources together and pass them in the constructor.